### PR TITLE
Dhg multithread

### PR DESCRIPTION
**Pull request by DaniDelHoyo**
Several bugs fixed 
More complete test
Multithread allowed

### CODE REVIEW

**Comment by DaniDelHoyo** [comment]
It might fix https://github.com/I2PC/scipion-em-xmipp/issues/426

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 1]
```diff
@@ -465,81 +472,52 @@ def _insertAllSteps(self):
...
+
+        elif trainedParams['posParticlesTrained'] >= toTrainSize and trainedParams['trainingPass'] != '':
+          lastTrainingPass = trainedParams['trainingPass']
+          self.retrievePreviousPassModel('', lastTrainingPass)
+          trainedParams['trainingPass'] = ''
+          self.saveTrainedParams(trainedParams)
+
+        #Prediction
+        if self.networkReadyToPredict() and self.cnnFree() and self.predictionsOn():
+            self.PREDICTING = True
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:503]: in other cases, you test also for `and if not in_this_state` (see line 484 and 485)

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 2]
```diff
@@ -635,20 +585,27 @@ def initializeStep(self):
...
+        if self._useNewModel() and self.skipTraining.get():
+          self.skipTraining.set(False)
+
         preprocessParamsFname= self._getExtraPath("preprocess_params.json")
         preprocParams= self.getPreProcParamsFromForm()
         with open(preprocessParamsFname, "w") as f:
           json.dump(preprocParams, f)
 
         #Initializing outputs
+        self.inpSampligRate = self._getInputMicrographs().getSamplingRate()
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:601]: extra `p` in `inpSampligRate`?

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 3]
```diff
@@ -1275,30 +1259,33 @@ def getAllCoordsInputMicrographs(self, shared=False):
...
           if inputMicrographs is None:
             raise ValueError("there are problems with your coordiantes, they do not have associated micrographs ")
           self.inputMicrographs = inputMicrographs
+        self.USING_INPUT_COORDS = False
       return self.inputMicrographs
 
     def _getBoxSize(self):
       '''Returns the box size of the input coordinates'''
       if not hasattr(self, "boxSize") or not self.boxSize:
+        self.waitFreeInputCoords()
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:1286]: this `waitFreeInputCoords` seems to be used quite a lot. Maybe we have a design issue?

File: xmipp3/protocols/protocol_screen_deepConsensus.py [hunk 4]
```diff
@@ -1441,12 +1390,16 @@ def allFree(self):
...
+      finished=True
       for coords in self.inputCoordinates:
         coords = coords.get()
         coords.loadAllProperties()
         if coords.isStreamOpen():
-          return False
-      return True
+          finished = False
+          break
+      self.USING_INPUT_COORDS = False
```
Comment (DStrelak) [inline xmipp3/protocols/protocol_screen_deepConsensus.py:1401]: this looks like side effect to me. Also why should I wait for coordinates (line 1393) when I'm checking if source protocols finished?

File: xmipp3/tests/test_protocol_deepConsensusTesting.py [hunk 5]
```diff
@@ -33,12 +33,14 @@
                             ProtCreateStreamData)
```
Comment (DStrelak) [inline xmipp3/tests/test_protocol_deepConsensusTesting.py:33]: why the renaming?

**Review by DStrelak — APPROVED** [review]

**Pull request closed without merging**